src/delta_debugging_bitmap.rs

```diff
@@ -0,0 +1,294 @@
+                self.ones_probability
+            );
+        }
+        let size = Uniform::new(self.size.start, self.size.end - 1).sample(runner.rng());
+        let iter = (0..size as u32).filter(|_| runner.rng().gen_bool(self.ones_probability));
+        let bitmap =
+            RoaringBitmap::from_sorted_iter(iter).map_err(|e| Reason::from(e.to_string()))?;
+        Ok(DeltaDebuggingBitmapValueTree::new(bitmap))
+    }
+}
+
+/// This struct holds the state needed for an implementation of the
+/// [delta debugging] algorithm on bitmaps.
+///
+/// Together with proptest's shrinking code and the `ValueTree` implementation,
+/// this shrinks bitmaps in O(log(n)) time to a single bit causing a failure,
+/// or shrinks bitmaps in O(n²) time to exactly the bits causing the failure
+/// (more precisely: a set of bits where removing a single one of them will make
+/// the failure go away).
+///
+/// [delta debugging]: https://www.st.cs.uni-saarland.de/papers/tse2002/tse2002.pdf
+#[derive(Debug)]
+pub struct DeltaDebuggingBitmapValueTree {
+    upper: RoaringBitmap,
+    split_into: u64,
+    split_index: u64,
+    complement: bool,
+}
+
+impl DeltaDebuggingBitmapValueTree {
+    pub fn new(bitmap: RoaringBitmap) -> Self {
+        // We initialize `split_index` to 1, so .current() returns
+        // `bitmap` initially.
+        Self {
+            upper: bitmap,
+            split_into: 1,
+            split_index: 0,
+            complement: false,
+        }
+    }
+}
+
+impl ValueTree for DeltaDebuggingBitmapValueTree {
+    type Value = RoaringBitmap;
+
+    fn current(&self) -> Self::Value {
+        let mut current = RoaringBitmap::new();
+        let num_items = self.upper.len();
+        let split_width = num_items / self.split_into;
+
+        let mut idx = 0;
+        for n in self.upper.iter() {
+            let split_start = split_width * self.split_index;
+            let split_end = split_start + split_width;
+            let in_current_split = split_start <= idx && idx < split_end;
+
+            if in_current_split ^ self.complement {
+                current.insert(n);
+            }
+
+            idx += 1;
+        }
+
+        current
+    }
+
+    fn simplify(&mut self) -> bool {
+        // This is called when the test case failed (as it should during shrinking!)
+
+        let num_elems = self.upper.len();
+        if num_elems == 0 || num_elems == 1 {
+            return false;
+        }
+
+        self.upper = self.current();
+
+        if self.complement {
+            // Successfully reduced to subset complement
+            self.split_into = max(self.split_into - 1, 2);
+            self.complement = false;
+        } else if self.split_into == 1 {
+            // We add a special case to test the empty set as early as possible
+            self.complement = true;
+        } else {
+            // Successfully reduced to subset
+            self.split_into = 2;
+            self.complement = false;
+        }
```

Comment:
Hmm. I'm generally a fan of representing state machines using `enum`s instead of booleans, but in this case it really complicates the implementation quite a bit. E.g. look at the `ValueTree::current` function.
With `Self` represented as:
```rs
pub enum DeltaDebuggingBitmapValueTree {
    CheckSubset {
        upper: RoaringBitmap,
        split_into: u64,
        split_index: u64,
    },
    CheckSubsetComplement {
        upper: RoaringBitmap,
        split_into: u64,
        split_index: u64,
    },
}
```

You end up needing to match as the very first thing in the function, to get access to `upper` and `split_into`.
The fact that it's checking the complement vs. not the complement is only needed in the innermost part of the function however.

I really think in this case it's better to represent it as is.

Reply:
There might have been a slight misunderstanding :)

It's totally fine to lift the enum fields used by all states into the parent struct.  Only the state-specific stuff should go into enums.  Same with `ReductionState` :)

Reply:
Ah okay - tackling some boolean blindness. I'm happy with doing that :)

Reply:
> Ah okay - tackling some boolean blindness. I'm happy with doing that :)

Exactly, that's the name for it 👍